app/feature/feature-movingflow/src/main/kotlin/com/hedvig/android/feature/movingflow/storage/MovingFlowStorage.kt

```diff
@@ -0,0 +1,52 @@
+package com.hedvig.android.feature.movingflow.storage
+
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.edit
+import androidx.datastore.preferences.core.stringPreferencesKey
+import com.hedvig.android.feature.movingflow.data.MovingFlowState
+import com.hedvig.android.logger.LogPriority
+import com.hedvig.android.logger.logcat
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.map
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+
+internal class MovingFlowStorage(
+  private val dataStore: DataStore<Preferences>,
+) {
+  fun getMovingFlowState(): Flow<MovingFlowState?> {
+    return dataStore.data.map { preferences ->
+      preferences[movingFlowStateKey]?.let { Json.decodeFromString<MovingFlowState>(it) }
+    }
+  }
+
+  suspend fun setMovingFlowState(movingFlowState: MovingFlowState) {
+    dataStore.edit { preferences ->
+      preferences[movingFlowStateKey] = Json.encodeToString(movingFlowState)
+    }
+  }
+
+  /**
+   * Does nothing if there was no moving flow state to edit.
+   * Returns the edited [MovingFlowState]
+   */
+  suspend inline fun editMovingFlowState(crossinline block: (MovingFlowState) -> MovingFlowState): MovingFlowState? {
+    dataStore.edit { preferences ->
+      val oldState = preferences[movingFlowStateKey]?.let { Json.decodeFromString<MovingFlowState>(it) }
+      if (oldState != null) {
+        val newState = block(oldState)
+        preferences[movingFlowStateKey] = Json.encodeToString(newState)
+      } else {
+        logcat(LogPriority.ERROR) { "Trying to `editMovingFlowState` a non-existing moving flow state" }
```

Comment:
should we mb return ErrorMessage in this case?

Reply:
If anything I was even thinking of throwing an Exception here, but I figured doing nothing and showing an Error would be okay.
I can not possibly imagine a scenario where we are in the screens where we are editing a moving flow, but without having created a moving flow in the first place in the start screen.
I suspect we are never going to see this log, and if we do it means something is very very off